### PR TITLE
perf(web): optimize compression streams

### PR DESCRIPTION
**Pull request by littledivy**
- [x] Use fast calls for `op_compression_new`.
- [x] Use `zlib-ng` instead of default `miniz_oxide` backend. Known to be much faster. https://github.com/rust-lang/flate2-rs#backends

~1.3x faster than Node's `gzip`/`gunzip` zlib **synchronous** APIs. 

```
# this patch
target/release/deno run -A --unstable cli/bench/gzip/compression.js
uncompressed size 889385
compressed size 874462

roundtrip
time 1850 ms rate 54
time 1847 ms rate 54
time 1852 ms rate 53
gzip
time 1521 ms rate 65
time 1522 ms rate 65
time 1520 ms rate 65
gunzip
time 299 ms rate 334
time 306 ms rate 326
time 300 ms rate 333
```

```
node cli/bench/gzip/compression.js
uncompressed size 889385
compressed size 871662

roundtrip
time 2398 ms rate 41
time 2395 ms rate 41
time 2381 ms rate 41
gzipSync
time 2072 ms rate 48
time 2078 ms rate 48
time 2072 ms rate 48
gunzipSync
time 313 ms rate 319
time 314 ms rate 318
time 313 ms rate 319
```

```
# main
deno run -A --unstable cli/bench/g
zip/compression.js
uncompressed size 889385
compressed size 868641

roundtrip
time 2238 ms rate 44
time 2254 ms rate 44
time 2222 ms rate 45
gzip
time 1882 ms rate 53
time 1881 ms rate 53
time 1883 ms rate 53
gunzip
time 332 ms rate 301
time 335 ms rate 298
time 334 ms rate 299
```

### CODE REVIEW

File: ext/web/compression.rs [hunk 1]
```diff
@@ -38,26 +38,24 @@ impl Resource for CompressionResource {
...
-#[op]
+#[op(fast)]
 pub fn op_compression_new(
   state: &mut OpState,
-  format: String,
+  format: u32,
   is_decoder: bool,
-) -> ResourceId {
+) -> u32 {
   let w = Vec::new();
```
Comment (aapoalas) [inline ext/web/compression.rs:47]: Why always the replacing of ResourceId with a plain u32?

File: ext/web/compression.rs [hunk 2]
```diff
@@ -38,26 +38,24 @@ impl Resource for CompressionResource {
...
+    (1, false) => {
       Inner::DeflateRawEncoder(DeflateEncoder::new(w, Compression::default()))
     }
-    ("gzip", true) => Inner::GzDecoder(GzDecoder::new(w)),
-    ("gzip", false) => {
-      Inner::GzEncoder(GzEncoder::new(w, Compression::default()))
-    }
+    (2, true) => Inner::GzDecoder(GzDecoder::new(w)),
+    (2, false) => Inner::GzEncoder(GzEncoder::new(w, Compression::default())),
     _ => unreachable!(),
```
Comment (aapoalas) [inline ext/web/compression.rs:59]: Shouldn't be a panic I think, but rather an error.

File: ext/web/compression.rs (hunk 1 above)
Comment (littledivy) [reply ext/web/compression.rs:47]: umm...`Result<ResourceId, _>` isn't working

**Comment by aapoalas** [comment]
Regarding the writing and out buffers I started wondering if it wouldn't be possible to do a lot of ZeroCopyBuf returning ops as pairs of ops where the first does the work and saves the return as a Resource, returning a ResourceId plus, using an out ptr, the required buffer size. The second call would then be called with a newly allocated buffer based on the out ptr received buffer size.

File: ext/web/compression.rs (hunk 1 above)
Comment (aapoalas) [reply ext/web/compression.rs:48]: Huh. Gotta take a look at that.

**Comment by littledivy** [comment]
> Regarding the writing and out buffers I started wondering if it wouldn't be possible to do a lot of ZeroCopyBuf returning ops as pairs of ops where the first does the work and saves the return as a Resource, returning a ResourceId plus, using an out ptr, the required buffer size. The second call would then be called with a newly allocated buffer based on the out ptr received buffer size.

That will be useful in other cases, yes.

When compressing, its hard to estimate size of the buffer which makes it very easy to lose performance by allocating way too much. When decompressing, its going to be larger than the input buffer and we have the same problem with estimation here. ZeroCopyBuf is probably the best in this situation.

**Comment by billywhizz** [comment]
> and saves the return as a Resource,

@aapoalas this is an interesting thought and similar to one i was having myself if i understand you correctly. i don't know enough about internals of Resources in Deno as yet but the pattern i have found most performant for this kind of thing is along these lines:

- allocate an arraybuffer in JS
- pass this arraybuffer into C++/Rust where pointer to and size of data is stored in a Map with an integer key that is returned to JS
- when performing any ops on this buffer, just pass in the integer key and have C++/Rust lookup the struct. The struct looks something like this and can be mapped onto another ArrayBuffer which JS side can read values from using TypedArray/DataView
```
struct rawBuffer {
  uint32_t len;
  int32_t read;
  int32_t written;
  uint32_t id;
  uint8_t* data;
};

std::map<int, rawBuffer*> buffers;
```
We could also store a persistent v8 handle on this struct to ensure the buffer is not moved or collected and expect consumer to explicitly release it when they are done (which is likely never as these will be buffers used repeatedly).

then, the op code becomes something like this, which is very performant as only passing an integer key from JS to C++/Rust, writing the results into internal struct and not returning any value.

```
void just::html::EscapeSIMDRaw2(const FunctionCallbackInfo<Value> &args) {
  just::html::rawBuffer* src = just::html::buffers[Local<Integer>::Cast(args[1])->Value()];
  src->written = args[0].As<String>()->WriteUtf8(args.GetIsolate(), (char*)src->data, src->len, &src->read, 
    v8::String::HINT_MANY_WRITES_EXPECTED | 
    v8::String::REPLACE_INVALID_UTF8 | 
    v8::String::NO_NULL_TERMINATION
  );
  just::html::rawBuffer* dest = just::html::buffers[Local<Integer>::Cast(args[2])->Value()];
  dest->written = hesc_escape_html(dest->data, src->data, src->written);
  args.GetReturnValue().Set(String::NewFromUtf8(args.GetIsolate(), (char*)dest->data, 
    NewStringType::kNormal, dest->written).ToLocalChecked());
}
```

of course, there is no error checking in this example, but it seems we can avoid that overhead now with fast calls and once the code is optimized by JIT we should be on fast path where we can bypass the expensive error checks.

**Comment by littledivy** [comment]
oh wait i misunderstood @aapoalas...makes sense now! yeah we should try this

**Comment by billywhizz** [comment]
> oh wait i misunderstood @aapoalas...makes sense now! yeah we should try this

what i was basically thinking is being able to allocate an arrabyffer as a resource and pass that around. then Rust side can have some certainty around properties of the buffer without lots of checks. should also help with data locality/cpu cache as reusing same memory repeatedly. there may be security implications to think about.

**Comment by aapoalas** [comment]
> what i was basically thinking is being able to allocate an arrabyffer as a resource and pass that around. then Rust side can have some certainty around properties of the buffer without lots of checks. should also help with data locality/cpu cache as reusing same memory repeatedly. there may be security implications to think about.

One issue I can see with this is that it is observable and somewhat spooky on a usage level.

### Observability
If an API returns an `ArrayBuffer` and that buffer is one of the "resource buffers", then its length will not be exactly the size that is needed to store the data, but whatever the resource buffer was allocated with. The only way around this is to make a sized copy of the buffer using `buffer.slice(0, dataByteLength)` but now we're making an extra copy.

Alternatively, if an API returns eg. `Uint8Array`, its size can be sized to the data size but the user can still observe the disparity in data size and buffer size by using `uint8Array.buffer` to get access to the backing `ArrayBuffer`.

### Spooky
If copies are avoided and the same `ArrayBuffer` is being reused and returned, then users have to just know that they don't really own the buffer they're given.

eg. This seems like a safe thing to do:
```ts
const someBuffer = Deno.readFileSync("./myFile.txt");
await mutateBufferData(someBuffer);
Deno.writeFileSync("./myFile.mutated.txt", someBuffer);
```
You receive a buffer, mutate its data and then write the result into a file. But what if during your `await` handling another `Deno.readFile` call happens to be made and it reuses the `ArrayBuffer` backing `someBuffer.buffer`? Suddenly your `someBuffer` contents will be something entirely different.

**Comment by billywhizz** [comment]
i did some testing locally on linux with zlib-ng and your example image and i don't see the drastic improvement you see even with zlib-ng compiled with --native flag to take advantage of all local CPU optimizations. it could be down to some processor extension you have that zlib-ng is able to take advantage of?

the png image is probably not a very good example to use i think as png is already compressed and you only shave a few bytes off the file by compressing it again. on the 128k zlib-ng shared library file i do see a dramatic improvement using zlib-ng over stock zlib and versus node.js on my system.

## uncompressed.png
Screenshot from 2022-09-25 19-26-58

## libz.so
(using zlib-ng compiled with --zlib-compat and --native flags)
Screenshot from 2022-09-25 19-30-31

## libz.so
(using zlib-ng compiled with --zlib-compat only - assumes existence of SSE on x86_64)
Screenshot from 2022-09-25 19-34-28

this is the little C wrapper i wrote which i call using Deno FFI.
```C
uint32_t zlib_deflate (uint8_t* src, uint32_t ssize, uint8_t* dest, uint32_t dsize) {
  z_stream* stream = (z_stream*)calloc(1, sizeof(z_stream));
  unsigned int compression = Z_DEFAULT_COMPRESSION;
  int windowbits = 31;
  deflateInit2(stream, compression, Z_DEFLATED, windowbits, Z_DEFAULT_MEMLEVEL, Z_DEFAULT_STRATEGY);
  stream->next_in = (Bytef*)src;
  stream->avail_in = ssize;
  stream->next_out = (Bytef*)dest;
  stream->avail_out = dsize;
  uint32_t avail_out = stream->avail_out;
  uint32_t flush = Z_FINISH;
  deflate(stream, flush);
  uint32_t written = avail_out - stream->avail_out;
  deflateEnd(stream);
  return written;
}
```

**Comment by billywhizz** [comment]
@aapoalas yes - i can think of lots of issues with this especially when doing async ops. i was thinking more of it as something we might be able to use internally when we are doing synchronous things like encoding/escaping/crypto. definitely needs some further discussion and experimentation, but the potential perf improvements make that worth doing i think?

File: ext/web/Cargo.toml [hunk 3]
```diff
@@ -18,7 +18,7 @@ async-trait.workspace = true
 base64-simd = "0.7"
 deno_core.workspace = true
 encoding_rs.workspace = true
-flate2.workspace = true
+flate2 = { version = "1.0.24", features = ["zlib-ng"], default-features = false }
```
Comment (bartlomieju) [inline ext/web/Cargo.toml:21]: Can this be moved to the root `Cargo.toml`?

File: tools/wpt/expectation.json [hunk 4]
```diff
@@ -4769,7 +4769,7 @@
         "text/html;a]=bar;b[=bar;c=bar (Blob/File)",
         "text/html;in]valid=\";asd=foo\";foo=bar (Blob/File)",
         "!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz/!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz;!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz=!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz (Blob/File)",
-        "x/x;x=\"\t !\\\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ\" (Blob/File)",
+        "x/x;x=\"\t !\\\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ\" (Blob/File)",
```
Comment (bartlomieju) [inline tools/wpt/expectation.json:4772]: These changes cause WPT failures

File: cli/bench/gzip/compression.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+function decompress(buffer) {
+  const ds = new DecompressionStream("gzip");
+  const writer = ds.writable.getWriter();
+  writer.write(buffer);
+  writer.close();
+  return (new Response(ds.readable).arrayBuffer());
+}
+
+const data = typeof Deno !== "undefined"
+  ? Deno.readFileSync("cli/bench/gzip/uncompressed.png")
```
Comment (bartlomieju) [inline cli/bench/gzip/compression.js:39]: Can you replace the current image with one of the images from the Deno blog post?

**Review by aapoalas — CHANGES_REQUESTED** [review]
Requires a rebase. @bartlomieju also had some good comments that should be addressed (especially WebIDL failures). Maybe this is also relevant w.r.t. the Node compat zlib work?

Anyway, this is one more of these nice perf PR's that would be nice to get merged.

**Comment by bartlomieju** [comment]
@littledivy @mmastrac should we continue to pursue this PR or close it in favor of the "fast streams" work that Matt is doing?

**Comment by bartlomieju** [comment]
I'm gonna close this PR for now for the lack of updates.

**Pull request closed without merging**